src/stores/chat.ts

```diff
@@ -0,0 +1,162 @@
+/**
+ * SPDX-FileCopyrightText: 2025 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+import type {
+	ChatMessage,
+	TokenMap,
+} from '../types/index.ts'
+
+import { defineStore } from 'pinia'
+import { reactive } from 'vue'
+import { useStore } from 'vuex'
+
+type ProcessChatBlocksOptions = {
+	/** if given, look for Set that has it */
+	mergeBy?: number
+}
+
+/**
+ * Check, if two sets intersect with each other
+ * Same complexity and result as !Set.prototype.isDisjointFrom()
+ */
+function checkIfIntersect(parentBlock: Set<number>, childBlock: Set<number>): boolean {
+	for (const id of childBlock) {
+		if (parentBlock.has(id)) {
+			return true
+		}
+	}
+	return false
+}
+
+/**
+ * Store for conversation chats
+ */
+export const useChatStore = defineStore('chat', () => {
+	const store = useStore()
+
+	const chatBlocks = reactive<TokenMap<Set<number>[]>>({})
+
+	/**
+	 * Returns list of messages, belonging to current context
+	 */
+	function getMessagesList(token: string): ChatMessage[] {
+		if (!store.state.messagesStore.messages[token] || !chatBlocks[token]) {
+			return []
+		}
+
+		// FIXME temporary show all messages from al blocks - no behaviour change
+		return Array.from(chatBlocks[token].flatMap((set) => Array.from(set)))
+			.sort((a, b) => a - b)
+			.reduce<ChatMessage[]>((acc, id) => {
+				const message = store.state.messagesStore.messages[token][id]
+				if (message) {
+					// If message is not found in the store, it's an invisible system or expired message
+					acc.push(message)
+				}
+				return acc
+			}, [])
+	}
+
+	/**
+	 * Populate chat blocks from given arrays of messages
+	 * If blocks already exist, try to extend them
+	 */
+	function processChatBlocks(token: string, messages: ChatMessage[], options?: ProcessChatBlocksOptions): void {
+		const newMessageIds = messages.map((message) => message.id)
+		const newMessageIdsSet = new Set(newMessageIds)
+
+		if (!chatBlocks[token]) {
+			// If no blocks exist, create a new one with the first message. First in array will be considered main block
+			chatBlocks[token] = [newMessageIdsSet]
+			return
+		}
+
+		if (options?.mergeBy) {
+			newMessageIdsSet.add(options.mergeBy)
+		}
+
+		if (!chatBlocks[token].some((parentBlock) => tryMergeBlocks(parentBlock, newMessageIdsSet))) {
+			chatBlocks[token].push(newMessageIdsSet)
```

Comment:
```suggestion
			if (!chatBlocks[token].some((parentBlock) => tryMergeBlocks(parentBlock, newMessageIdsSet))) {
			let i = chatBlocks[token].length - 1
			while (i >= 0 && Math.max(...Array.from(chatBlocks[token][i])) >= Math.max(...Array.from(newMessageIdsSet))) {
				i--
			}
			chatBlocks[token].splice(i + 1, 0, newMessageIdsSet)
		}
```

I remember this logic from high school but feel free to adjust to modern way :') 

If we directly add the item in the right position, we will have the array always sorted and no need to iterate over all of it each time 

~~mergeAndSortChatBlocks(token)~~

Reply:
If dropping postcheck, this needs adjustments for merging `A + new + B` wherever it's inserted

Reply:
Whenever we try to insert 'new', it will be when merged if needed `tryMergeBlocks(parentBlock, newMessageIdsSet)` or inserted in a sorted way